### PR TITLE
ci: automate release to OpenWRT

### PR DESCRIPTION
**Pull request by mad-it**
⚠️ WIP ⚠️  

An attempt to automate release to OpenWRT. Would love to hear your thoughts.

The idea behind this is to automate creating of PR towards OpenWRT with the `Makefile` updated with the version number + updated hash.

Resolves https://github.com/nextdns/nextdns/issues/182

### CODE REVIEW

**Comment by rs** [comment]
You need the main repo to get the build tools, then:
* add packages to it
* make config to enable nextdns as a module
* update the version in the makefile
* make packages/nextdns/download to fetch the tarball
* make packages/nextdns/prepare FIXUP=1 to update the checksum in the Makefile
* ideally build the package to ensure it works
* create a PR for master and another for the 19.07.1

I guess we could skip most of that if we extract then logic to create the checksum.

**Comment by mad-it** [comment]
The only thing we need to automate is the creation of the following PR: https://github.com/openwrt/packages/pull/11411

This means 2 things need to be updated:

- PKG_MIRROR_HASH
- PKG_VERSION

From the information I could gather, PKG_MIRROR_HASH should be a SHA256 of the git repo. Im assuming this is just cloning the entire repository on of the version tag and hashing this. This hash is used as a caching mechanism on their side to not retrieve the sources from upstream each time.
Source: https://forum.openwrt.org/t/what-pkg-mirror-hash-or-pkg-hash-means/1602

As for PKG_VERSION, this is pretty trivial. 

This should be enough to automate releases towards OpenWRT right?

**Comment by rs** [comment]
We could certain generate the hash without doing all that, I haven't check how it's done but it should be simple enough. This would publish a non tested package tho, but this package is simple enough so that should be fine.

**Comment by mad-it** [comment]
After some nice digging, if found out how they check the hash(I hope). https://github.com/openwrt/openwrt/blob/master/scripts/dl_github_archive.py#L250

It looks like they just download the Tarball from GitHub and run a checksum on that file. I think they untar it and re-tar the file in case `PKG_SOURCE_SUBDIR` is set. But in the case for NextDNS this is not the case so I'm guessing simple getting the checksum of the `tar.gz` of the version we want to release should suffice. 

I had to trace everything back to https://github.com/openwrt/openwrt/blob/09c428ec6bca27bdb31e2f6fef85b07787cead30/include/download.mk#L186

I hope you reach the same conclusion as I have 🙈

**Comment by rs** [comment]
Looks like they tar but don't gzip on repack. Am I reading wrong?

**Comment by rs** [comment]
@hnyman: perhaps you can help us on finding the best solution?

**Comment by mad-it** [comment]
After some reverse engineering of the python script and loosing a bunch of hair in the process, I was able to generate the correct hash in a single bash command.

You can even validate these hashes with the previous releases if you would like. Just change the `VERSION` env variable and the hash should be printed out. 
If you think this is enough to move forward Ill finish the PR and we can automate the release.

```bash
export VERSION=1.4.23 && \
export MTIME=$(curl -s https://api.github.com/repos/nextdns/nextdns/commits/v$VERSION | python3 -c "import sys, json, datetime, calendar; print(calendar.timegm(datetime.datetime.strptime(json.load(sys.stdin)['commit']['committer']['date'], '%Y-%m-%dT%H:%M:%SZ').timetuple()))") && \
umask 022 && \
curl -Ls https://api.github.com/repos/nextdns/nextdns/tarball/v$VERSION | tar -xz --no-same-permissions -C /tmp && mv nextdns-* nextdns-$VERSION &&  XZ_OPT=-7e tar --numeric-owner --owner=0 --group=0 --sort=name --mtime=@$MTIME -J -C /tmp  -cf /tmp/nextdns-$VERSION.tar.xz nextdns-$VERSION && \
sha256sum nextdns-$VERSION.tar.xz | awk '{print $1}' && \
rm -r nextdns-$VERSION.tar.xz nextdns-$VERSION
```

**Comment by mad-it** [comment]
Alright, Its ready. Tested it against a fork in my repository. 
The 2 following links are the result:
GitHub actions run: https://github.com/mad-it/nextdns/runs/630692323
Resulting pull request of that run: https://github.com/mad-it/packages/pull/5

TODO:
- [ ] Changes repositories from my local fork to the correct repo.

**Comment by rs** [comment]
That's awesome, I'll review that tonight. Thx!

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -9,26 +9,57 @@ jobs:
...
+          sed -i 's/PKG_VERSION:=.*/PKG_VERSION:=${{ env.VERSION }}/' net/nextdns/Makefile
+          sed -i 's/PKG_MIRROR_HASH:=.*/PKG_MIRROR_HASH:=${{ env.OPENWRT_HASH }}/' net/nextdns/Makefile
+      - name: Create pull request towards OpenWRT
+        uses: peter-evans/create-pull-request@v2
+        with:
+          token: ${{ secrets.RELEASE_GH_TOKEN }}
+          commit-message: "nextdns: Update to version ${{ env.VERSION }}"
+          title: "nextdns: Update to version ${{ env.VERSION }}"
+          branch: "rs-nextdns-${{ env.VERSION }}"
+          body: "Update to version ${{ env.VERSION }}"
```
Comment (rs) [inline .github/workflows/release.yml:65]: We miss another PR rebased on the 19.07.1 branch.

Comment (mad-it) [reply .github/workflows/release.yml:65]: There is no 19.07.1 branch in the packages repository of OpenWRT. There is only openwrt-19.07. Did you mean this branch by any chance?

Comment (rs) [reply .github/workflows/release.yml:65]: Yes sorry.

**Comment by hnyman** [comment]
Just a reference, Weblate translation tool pushed this kind of automatic Github PR about new translations to LuCI, the OpenWrt GUI.   

Weblate re-uses the same branch for the PRs, so that if the previous PR has not yet been merged, the new commits are added to the same PR and the commits are actually squashed together. (That is handy in case of translations, where individual translators do translation in small pieces and weblate gradually pushed their work to the PR, often sevral times per day. That prevents conflicts, if there is new work to be committed before the previous PR has been merged. I am not sure if that is relevant to you, but just wanted to point out a reference where this kind of automatic outside PR authoring already happend in the OpenWrt universe.)

**Comment by hnyman** [comment]
Regarding the HASH variables in OpenWrt downloads:

With all download versions, a copy will automatically placed (by the buildbot) into sources.openwrt.org in order to avoid bothering upstream too often. To enable verification, there are two different main methods:

If the download from upstream is
* a direct upstream version (PKG_VERSION) based published  .tar.xz or .tar.gz downloaded from upstream Github releases, Makefile has PKG_HASH that is a direct SHA256 of the upstream tar.xz
   Example: https://github.com/openwrt/packages/blob/master/utils/nano/Makefile 
* a git commit hash (PKG_SOURCE_VERSION) based local git clone (PKG_SOURCE_PROTO:=git, which gets then locally gzipped after download, there will be PKG_MIRROR_HASH matching the SHA256 calcualted from the locally cloned stuff, so that the copy placed to sources.openwrt.org can be easily verified next time. 
 example: https://github.com/openwrt/packages/blob/ffdbc7a637df8e3f39dd6916405f78840edccfb4/utils/mktorrent/Makefile 

If you have a downloadable .tar.gz in your Github releases, you only need the simple SHA256 calculated PKG_HASH of that.

**Comment by rs** [comment]
Thank you @hnyman, that's very useful.

**Comment by mad-it** [comment]
I totally forgot to add the other branch. This is now fixed. 

Here is the action run: https://github.com/mad-it/nextdns/runs/632365702
Here is PR created to master https://github.com/mad-it/packages/pull/7
Here is PR created to 19.07 branch https://github.com/mad-it/packages/pull/8

One thing to note in the created PR's is that the fork used(in these test cases my fork) is out of date so the PR's look a bit bigger then they should. The reason for this is that the option `request-on-parent: true` is commented out. Once this option is enabled it will create a PR to the upstream repository which is up to date.

TODO:
- [ ] Changes repositories from my local fork to the correct repo.
- [ ] Enable `request-on-parent: true`

**Comment by rs** [comment]
LGTM, please point to official repo and I will merge. Thanks a tone!

**Comment by mad-it** [comment]
@rs Its done. Just validated that the fork PR is working. 
Here is a PR created from a forked account https://github.com/mad-it/packages/pull/9.

Only remaining thing is if you want the commit to reflect your name and email by configuring `comitter` and `author` as described here https://github.com/marketplace/actions/create-pull-request#action-inputs

Those are more cosmetics and I will leave that up to you :)

**Comment by rs** [comment]
Seems like the tarball is not available during the build: https://github.com/nextdns/nextdns/runs/651575850?check_suite_focus=true

**Comment by rs** [comment]
The path was wrong. New issue on openWRT CI side:

```
Author name (github-actions[bot]) need to be your real name 'firstname lastname'
Commit subject line seems ok (nextdns: Update to version 1.5.6)
Signed-off-by is missing or doesn't match author (should be 'Signed-off-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>')
```

**Comment by mad-it** [comment]
Thanks for fixing the path. 

https://github.com/nextdns/nextdns/pull/203 should resolve the checks on OpenWRT side.

**Pull request merged**